tokenizer.go

```diff
@@ -316,24 +370,61 @@ func errorResponse(err error) *http.Response {
 	}
 }
 
-func forceTLSDialer(network, addr string) (net.Conn, error) {
-	switch network {
-	case "tcp", "tcp4", "tcp6":
-	default:
-		return nil, fmt.Errorf("%w: dialing network %s not supported", ErrBadRequest, network)
+// dialFunc returns a function for dialing network addresses. Ours does a few
+// special things.
+//   - It denies connections to the given set of IP addresses. This is to prevent
+//     circular requests back to tokenizer. Invalid IPs are ignored.
+//   - It rejects requests for TLS upstreams. We need to see/modify requests, so
+//     our proxy can't do passthrough TLS.
+//   - It forces the upstream connection to be TLS. We want the actual upstream
+//     connection to be over TLS because security.
+func dialFunc(badAddrs []string) func(string, string) (net.Conn, error) {
+	netDialer := net.Dialer{}
+
+	baMap := map[string]bool{}
+	for _, ba := range badAddrs {
+		if ip := net.ParseIP(ba); ip != nil {
+			baMap[ip.String()] = true
+		}
 	}
 
-	hostname, port, _ := strings.Cut(addr, ":")
-	if hostname == "" {
-		return nil, fmt.Errorf("%w: attempt to dial without host: %q", ErrBadRequest, addr)
-	}
-	switch port {
-	case "443":
-		return nil, fmt.Errorf("%w: proxied request must be HTTP", ErrBadRequest)
-	case "80", "":
-		port = "443"
+	if len(baMap) != 0 {
+		netDialer.ControlContext = func(ctx context.Context, network, address string, c syscall.RawConn) error {
+			h, _, err := net.SplitHostPort(address)
+			if err != nil {
+				return err
+			}
+
+			switch ip := net.ParseIP(h); {
+			case ip == nil:
+				return fmt.Errorf("bad ip: %s", address)
+			case baMap[ip.String()]:
```

Comment:
This is where we catch attempts to send requests to DNS names controlled by someone else that point at our static IPs.